staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -97,6 +111,40 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
 	return resp, ok, err
 }
 
-func keyFunc(auds []string, token string) string {
-	return fmt.Sprintf("%#v|%v", auds, token)
+// keyFunc generates a string hash based on the inputs.
+// This lowers the memory requirement of the cache and keeps tokens out of memory.
+func keyFunc(randomCachePrefix []byte, auds []string, token string) string {
+	// allocate a single buffer based on the longest string
+	maxLength := len(token)
+	for _, aud := range auds {
+		if length := len(aud); length > maxLength {
+			maxLength = length
+		}
+	}
+	b := make([]byte, 4+maxLength) // include extra bytes to store the length
+
+	h := sha256.New()
+
+	if _, err := h.Write(randomCachePrefix); err != nil {
+		panic(err) // Write() on hash never fails
+	}
+
+	hashLengthPrefixedString(h, b, token)
+	for _, aud := range auds {
+		hashLengthPrefixedString(h, b, aud)
+	}
+
+	return string(h.Sum(nil)) // skip base64 encoding to save an allocation
+}
+
+// hashLengthPrefixedString hashes s with a length prefix to prevent ambiguities, i.e. "xy" + "z" == "x" + "yz"
+// the length of b is assumed to be greater than or equal to the length of s + 4 (b is mutated by this function)
+func hashLengthPrefixedString(h hash.Hash, b []byte, s string) {
+	binary.BigEndian.PutUint32(b, uint32(len(s)))
+	if copy(b[4:], s) != len(s) {
+		panic("invalid scratch buffer")
+	}
+	if _, err := h.Write(b[:4+len(s)]); err != nil {
```

Comment:
There SHOULD be a fast way to write a string. https://golang.org/pkg/io/#WriteString comes close but sha256 doesn't support WriteString, so that's an alloc.

To get rid of this alloc I guess you'll have to do https://github.com/golang/go/issues/25484#issuecomment-391415660 :/